From eb9fc9243df085b4324e9f71e3f6d69059924be7 Mon Sep 17 00:00:00 2001
Date: Fri, 19 Jun 2015 13:53:22 +0200
Subject: [PATCH] ToChildBlockJoinQuery score calculation bugfix

Don't skip score calculation if first child is not in acceptDocs.
---
 .../org/apache/lucene/search/join/ToChildBlockJoinQuery.java | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/lucene/join/src/java/org/apache/lucene/search/join/ToChildBlockJoinQuery.java b/lucene/join/src/java/org/apache/lucene/search/join/ToChildBlockJoinQuery.java
index 2eaf1bf..3bb3ffb 100644
--- a/lucene/join/src/java/org/apache/lucene/search/join/ToChildBlockJoinQuery.java
+++ b/lucene/join/src/java/org/apache/lucene/search/join/ToChildBlockJoinQuery.java
@@ -217,7 +217,17 @@ public class ToChildBlockJoinQuery extends Query {
             }
 
             if (acceptDocs != null && !acceptDocs.get(childDoc)) {
-              continue nextChildDoc;
+              // find the first child that is accepted
+              while (true) {
+                if (childDoc+1 < parentDoc) {
+                  childDoc++;
+                  if (acceptDocs.get(childDoc))
+                    break;
+                } else {
+                  // no child for this parent doc matches acceptDocs
+                  continue nextChildDoc;
+                }
+              }
             }
 
             if (childDoc < parentDoc) {
-- 
1.9.5.github.0

